rust/candid/src/de.rs

```diff
@@ -847,30 +847,41 @@ impl<'de, 'a> de::VariantAccess<'de> for Compound<'a, 'de> {
 
 /// Allow decoding of any sized argument.
 pub trait ArgumentDecoder<'a>: Sized {
+    /// The types returned when decoding [Self] using [decode_one].
+    type FirstArg;
+
     /// Decodes a value of type [Self], modifying the deserializer (values are consumed).
     fn decode(de: &mut IDLDeserialize<'a>) -> Result<Self>;
 }
 
 /// Decode an empty tuple.
 impl<'a> ArgumentDecoder<'a> for () {
+    /// There is loss of type information when using `(())`, but that's expected and an
+    /// edge case.
+    type FirstArg = ();
```

Comment:
`()` means zero argument. So `FirstArg` is either `None`, `!`, or `candid::Empty`.